getline.ts/src/blockchain.ts

```diff
@@ -0,0 +1,237 @@
+                        }
+                    });
+                });
+            }, 5000);
+        });
+    }
+
+    public async deploy(contractName: string, ...params: Array<any>): Promise<Contract> {
+        let abi = await this.metabackend.getABI(contractName);
+        let contract = this.web3.eth.contract(abi);
+        let bytecode = await this.metabackend.getBytecode(contractName);
+
+        return new Promise<Contract>((resolve, reject)=>{
+            // TODO(q3k): Typify Web3.Contract.new
+            let contractAny: any = contract;
+
+            let opts = {
+                from: this.web3.eth.coinbase,
+                gas: 4500000,
+                data: bytecode,
+            };
+            console.log("getline.ts: deploying contract " + contractName);
+            console.log("getline.ts: deploying bytecode " + bytecode.substring(0, 64) + "...");
+            console.log("getline.ts:    with parameters " + params);
+
+            let confirmed = false;
+            let instance = contractAny.new(...params, opts, (err, c) =>{
+                if (err) {
+                    console.log("getline.ts: deployment failed: " + err.stack);
+                    reject(new Error("deployment failed: " + err));
+                    return;
+                }
+
+                // Make sure this callback gets fired only once...
+                if (confirmed) {
+                    return;
+                }
+                confirmed = true;
+
+                console.log("getline.ts: deploying...");
+                this.waitTxReceipt(c).then((address)=>{
+                    console.log("getline.ts: deployed at " + address.ascii);
+                    let deployed = this.web3.eth.contract(abi).at(address.ascii);
+                    resolve(new Contract(this, deployed));
+                }).catch(reject);
+            });
+        });
+    }
+
+    public async existing(name: string, address: Address): Promise<Contract> {
+        let abi = await this.metabackend.getABI(name);
+        let deployed = this.web3.eth.contract(abi).at(address.ascii);
+        return new Contract(this, deployed);
+    }
+
+    public async coinbase(): Promise<Address> {
+        return new Address(this, this.web3.eth.accounts[0]);
+    }
+
+    public currentBlock(): Promise<BigNumber> {
+        // TODO(q3k): Cache this?
+        return new Promise<BigNumber>((result, reject)=>{
+            this.web3.eth.getBlockNumber((err, block: number)=>{
+                if (err) {
+                    reject(err);
+                    return;
+                }
+                result(new BigNumber(block));
+            });
+        });
+    }
+
+    public blockToTime(current: BigNumber, block: BigNumber): moment.Moment {
+        if (this.network != "4") {
+            throw new Error("getline.ts only supports rinkeby chains");
+        }
+        let secondsPerBlock = new BigNumber(15);
```

Comment:
Future idea: Calculate last/average block time

Reply:
But useless anyways since we're moving to timestamps

Reply:
Ack.